Support setting arguments of pretraining by a config file

seopbo opened the pull request:
- config 파일하나로 pretrain_language_model.py에 실행에 필요한
  arguments를 전달함.
- nested dict 처리를 위한 Omegaconf library 추가
- `CONFIG_MAPPING`을 활용하여 class 생성자 호출

Refs: #16

aisolab commented:
현 시점으로서는 저는 LGTM 입니다.

bzantium commented:
vocab size를 default로 50257로 해주셨는데 그러면 configs/tokenizer/model/roberta.yaml 에서도 수정부탁드립니다!

bzantium commented:
[cls] [sep] token을 추가하더라도 보통 512로 하는게 맞지 않나요?

monologg commented:
@bzantium 

https://github.com/pytorch/fairseq/issues/1187

기록차원에서 링크 남겨 놓겠습니다:)

monologg commented on src/roberta/roberta_dataset_builder.py:
```diff
@@ -0,0 +1,67 @@
+from ..utils import list_dict_to_dict_list
+
+
+class RoBERTaDatasetBuilder(DatasetBuilder):
```
@soeque1 

마이너한 부분이긴 한데, `RoBERTa` 명칭을 `Roberta` 로 통일하는 것에 대해서 어떻게 생각하시나요?

이게 `transformers` 가 `RobertaForXXX` 로 사용하다보니, 이런 부분이 다르면 애매한 경우들이 종종 있더라고요ㅠㅠ

soeque1 replied:
@monologg 좋습니다!!

monologg commented on src/roberta/roberta_dataset_builder.py:
```diff
@@ -0,0 +1,67 @@
+        )["input_ids"]
+
+        for doc in docs:
+            doc = [self.tokenizer.cls_token_id] + doc
+            doc += [self.tokenizer.sep_token_id]
```
민호님께서 얘기해주신대로 `_create_example` 안에서 처리해주는 게 맞는것 같네요!

monologg commented on src/roberta/roberta_dataset_builder.py:
```diff
@@ -0,0 +1,67 @@
+        return None
+
+    def _create_example(self) -> transformers.BatchEncoding:
+        chunk = self.buffer[: self.max_length]
```
Q. 지금 구조에서는 context가 다른 문서가 이어붙여지는 건 상관없이, 무조건 full로 이어붙이는 방식인거죠?

soeque1 replied:
넵~

monologg commented on configs/pretrain/runner/roberta-small.yaml:
```diff
@@ -0,0 +1,31 @@
+  gradient_checkpointing: false
+  return_dict: true
+  bos_token_id: 0
+  eos_token_id: 0
```
`eos_token_id: 2` 로 바꿔야 할 것 같습니다!

monologg replied:
https://github.com/lassl/lassl/blob/main/configs/tokenizer/model/roberta.yaml#L18

지금 보니까 special token 순서가 원래 `roberta-base`와 상이하네요

monologg replied:
https://github.com/lassl/lassl/issues/28

soeque1 replied:
버그입니다 ㅋㅋ gpt2 복붙하다가 놓쳤네요 ~ ㅠ

bzantium commented on src/roberta/roberta_dataset_builder.py:
```diff
@@ -0,0 +1,67 @@
+    def _create_example(self) -> transformers.BatchEncoding:
+        chunk = self.buffer[: self.max_length]
+
+        example = self.tokenizer.prepare_for_model(
+            chunk, padding=False, truncation=False, return_attention_mask=True, return_token_type_ids=False,
+            add_special_tokens=False
```
대화 나눴던 것처럼 special_tokens를 뱉어주는 것으로 변경 요청드립니다!

soeque1 replied:
네 수정했습니당

bzantium commented on src/roberta/roberta_dataset_builder.py:
```diff
@@ -0,0 +1,67 @@
+            {
+                "input_ids": datasets.Sequence(datasets.Value("int32")),
+                "attention_mask": datasets.Sequence(datasets.Value("uint8")),
+                "position_ids": datasets.Sequence(datasets.Value("int32")),
```
이 부분도 "special_tokens" 추가 요망!

aisolab replied:
@bzantium mlm계열의 경우는 serialize 단계에서 `special_tokens_mask`를 받고, runner에서 사용하는 collator가 `[CLS]`와 `[SEP]`는 `special_tokens_mask`정보를 참고하여, `[CLS]`, `[SEP]`에 해당하는 걸 `[MASK]`하지않는 형태가 되겠죠?

bzantium replied:
네 맞습니다! 아래 보시면 해당 부분 prob을 0을 줘서 masking 못하게 하네요 :)
https://github.com/huggingface/transformers/blob/534f6eb9f1f8910a4912ccccd79f1f974c625168/examples/research_projects/performer/run_mlm_performer.py#L247

soeque1 replied:
네 ~ 수정했습니다

bzantium commented on src/roberta/roberta_dataset_builder.py:
```diff
@@ -0,0 +1,69 @@
+
+    def _create_example(self) -> transformers.BatchEncoding:
+        chunk = self.buffer[: self.max_length - 1]
+        chunk = [self.tokenizer.cls_token_id] + chunk
```
```suggestion
        chunk.insert(0, self.tokenizer.cls_token_id)
```
로 하는 것과 연산 차이가 없을까요?

soeque1 replied:
기존 방식은 리스트를 각각 만들어서 합치는 방식이라 더 개선 될 것 같아요.
실험도 그렇구요

```
import time

def main():
    test_vec = list(range(0, 1000000))
    s = time.time()
    test_vec = [-1] + test_vec
    print(time.time() - s)

    test_vec = list(range(0, 1000000))
    s = time.time()
    test_vec.insert(0, -1)
    print(time.time() - s)


if __name__ == '__main__':
    main()
```

bzantium reviewed (approved):
LGTM! 고생하셨습니다 :)

monologg reviewed (approved):
LGTM!

soeque1 commented:
close #18

Commit:
[Git] feat: Append ignore pycharm idea dir
- PyCharm IDEA 디렉토리 ignore

Refs: #16

Commit:
[Python] feat: Append ignore pycharm idea dir
- PyCharm IDEA 디렉토리 ignore

Refs: #16

Commit:
Merge branch 'feature/#16' of https://github.com/lassl/lassl into feature/#16

Commit:
Merge branch 'main' into feature/#16

seopbo committed:
[mixed] refactor: Configs for a pretrain script
- config 파일하나로 pretrain_language_model.py에 실행에 필요한
  arguments를 전달함.
- nested dict 처리를 위한 Omegaconf library 추가

Refs: #16

monologg reviewed (approved):
LGTM

iron-ij reviewed (commented):
sample 수준의 config 파일이라도 같이 포함되어야 할 것 같습니다. 아까 만드셨던 거라도 더해주시면 좋을 것 같아요~!!

seopbo commented:
> sample 수준의 config 파일이라도 같이 포함되어야 할 것 같습니다. 아까 만드셨던 거라도 더해주시면 좋을 것 같아요~!!

@iron-ij 아 그거 추가되어있을텐데요??

monologg committed:
[Git] Update .gitignore

seopbo committed:
[mixed] refactor: Replace *.json with *.yaml
- config를 yaml파일로 세팅

Refs: #16

seopbo committed:
Merge branch 'feature/#16' of github.com:lassl/lassl into feature/#16

seopbo commented:
> > sample 수준의 config 파일이라도 같이 포함되어야 할 것 같습니다. 아까 만드셨던 거라도 더해주시면 좋을 것 같아요~!!
> 
> @iron-ij 아 그거 추가되어있을텐데요??

없는 거 확인하고 추가했습니다.

bzantium reviewed (commented):
model_type의 가능 범주를 알려주는 조건절 명령어 추가

seopbo committed:
Update pretrain_language_model.py

Co-authored-by: Minho Ryu <ryumin93@gmail.com>

bzantium reviewed (approved):
LGTM :)
